streams/src/main/java/org/apache/kafka/streams/state/internals/metrics/RocksDBMetricsRecorder.java

```diff
@@ -87,47 +88,113 @@ public void addStatistics(final String storeName,
+                memtableBytesFlushed += statistics.getTickerCount(TickerType.FLUSH_WRITE_BYTES);
+                memtableHits += statistics.getAndResetTickerCount(TickerType.MEMTABLE_HIT);
+                memtableMisses += statistics.getAndResetTickerCount(TickerType.MEMTABLE_MISS);
+                blockCacheDataHits += statistics.getAndResetTickerCount(TickerType.BLOCK_CACHE_DATA_HIT);
+                blockCacheDataMisses += statistics.getAndResetTickerCount(TickerType.BLOCK_CACHE_DATA_MISS);
+                blockCacheIndexHits += statistics.getAndResetTickerCount(TickerType.BLOCK_CACHE_INDEX_HIT);
+                blockCacheIndexMisses += statistics.getAndResetTickerCount(TickerType.BLOCK_CACHE_INDEX_MISS);
+                blockCacheFilterHits += statistics.getAndResetTickerCount(TickerType.BLOCK_CACHE_FILTER_HIT);
+                blockCacheFilterMisses += statistics.getAndResetTickerCount(TickerType.BLOCK_CACHE_FILTER_MISS);
+                writeStallDuration += statistics.getTickerCount(TickerType.STALL_MICROS);
+                bytesWrittenDuringCompaction += statistics.getTickerCount(TickerType.COMPACT_WRITE_BYTES);
+                bytesReadDuringCompaction += statistics.getTickerCount(TickerType.COMPACT_READ_BYTES);
+                numberOfOpenFiles += statistics.getTickerCount(TickerType.NO_FILE_OPENS)
+                    - statistics.getTickerCount(TickerType.NO_FILE_CLOSES);
+                numberOfFileErrors += statistics.getTickerCount(TickerType.NO_FILE_ERRORS);
+            }
+        }
+        bytesWrittenToDatabaseSensor.record(bytesWrittenToDatabase);
+        bytesReadFromDatabaseSensor.record(bytesReadFromDatabase);
+        memtableBytesFlushedSensor.record(memtableBytesFlushed);
+        memtableHitRatioSensor.record(computeHitRatio(memtableHits, memtableMisses));
+        blockCacheDataHitRatioSensor.record(computeHitRatio(blockCacheDataHits, blockCacheDataMisses));
+        blockCacheIndexHitRatioSensor.record(computeHitRatio(blockCacheIndexHits, blockCacheIndexMisses));
+        blockCacheFilterHitRatioSensor.record(computeHitRatio(blockCacheFilterHits, blockCacheFilterMisses));
+        writeStallDurationSensor.record(writeStallDuration);
+        bytesWrittenDuringCompactionSensor.record(bytesWrittenDuringCompaction);
+        bytesReadDuringCompactionSensor.record(bytesReadDuringCompaction);
+        numberOfOpenFilesSensor.record(numberOfOpenFiles);
+        numberOfFileErrorsSensor.record(numberOfFileErrors);
+    }
+
+    private double computeHitRatio(final long hits, final long misses) {
+        if (hits == 0) {
+            return 0;
+        }
+        return (double) hits / (hits + misses);
     }
 
     public void close() {
+        stopRecording = true;
         for (final Statistics statistics : statisticsToRecord.values()) {
             statistics.close();
         }
         statisticsToRecord.clear();
+        log.debug("Closed", storeName);
```

Comment:
Should we check for the thread has eventually joined, say when the streams instance is closed? With 10min we should not block on waiting, but also we should make sure we are not accidentally leaking threads.

Reply:
When the last `Statistics` object is removed the thread is interrupted, stopped, and the removing thread waits for the thread to die. That should ensure no leak. Please review that code with special care.